staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -276,18 +277,33 @@ func RunTestWatchError(ctx context.Context, t *testing.T, store InterfaceWithPre
 		t.Fatalf("GuaranteedUpdate failed: %v", err)
 	}
 
-	// Now trigger watch error by injecting failing transformer.
-	revertTransformer := store.UpdatePrefixTransformer(
-		func(previousTransformer *PrefixTransformer) value.Transformer {
-			return &failingTransformer{}
-		})
-	defer revertTransformer()
+	// If we are testing against etcd, we need to deliberately trigger a watch error.
+	// This is not needed while testing with cacher since cacher will emit a resource
+	// expired error itself.
+	if !cacher {
+		// Now trigger watch error by injecting failing transformer.
+		revertTransformer := store.UpdatePrefixTransformer(
+			func(previousTransformer *PrefixTransformer) value.Transformer {
+				return &failingTransformer{err: rpctypes.ErrCompacted}
```

Comment:
That's extremly hacky to me.

Why do we need it? Can't we just configure test-etcd to set it with a small compaction period (like 5s or sth) and wait for the item to be compacted?